### PR TITLE
Add ApplyBuilder syntax

### PR DESCRIPTION
**Pull request by stephenjudkins**
This is done mostly through macros + quasiquotes. IMO, it's surprisingly readable.
This approach might work for more boilerplate avoidance and abstractions over arity.

### CODE REVIEW

**Comment by stephenjudkins** [comment]
See https://github.com/non/cats/issues/162

**Comment by stephenjudkins** [comment]
See https://gist.github.com/stephenjudkins/5ed917b547aa3782e04e for verification. Where should at automated test like this go? This would be very hard to test in a property-based way, but a simple spec verifying that the syntax works would be helpful

**Comment by non** [comment]
Looks like this fell foul of the authoritarian ScalaStyle:

```
[error] /home/travis/build/non/cats/macros/src/main/scala/cats/macros/apply.scala:68:6: Class name does not match the regular expression '[A-Z][A-Za-z]*'
```

**Comment by mpilquist** [comment]
I think it will fail under 2.10 too -- things like `whitebox` and `TypeName`.

**Comment by wedens** [comment]
@mpilquist cats already supports only 2.11 https://github.com/non/cats/pull/7

**Comment by mpilquist** [comment]
@wedens Travis ported Simulacrum to 2.10 shortly after that issue was written. I haven't released his work yet, but I will this week at some point. I think I'm going to release a 0.3.0 without binary type constructor support and then a 0.4.0 with binary type constructor support.

Anyway, I think the plan is to support 2.10. Doesn't have to be now of course, but might be better to avoid revisiting all the macros when we add 2.10.

**Comment by travisbrown** [comment]
If we're using macros for this, it wouldn't be too hard to add 2.10 support later (although I think avoiding macro bundles for now would be ideal).

Is plain old codegen off the table, though? As much as I like macro annotations, the benefit here isn't as clear as for the Simulacrum cases, etc.

**Comment by travisbrown** [comment]
Also, do we think `ApplicativeBuilder` will end up private eventually? If these aren't going to be Scaladoc'd, it might be a little cleaner to use `fresh` or identifiers with a numeric suffix instead of the alphabet.

File: core/src/main/scala/cats/ApplyBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package cats
```
Comment (travisbrown) [inline core/src/main/scala/cats/ApplyBuilder.scala]: Shouldn't this be in `syntax`?

**Comment by ceedubs** [comment]
It sounds like the consensus was to go with the approach taken in #176. Since that's been merged, I'm going to close this ticket out. Thanks for putting it together, @stephenjudkins!

**Pull request closed without merging**